File: 948-bag-of-tokens/948-bag-of-tokens.java
class Solution {
    public int bagOfTokensScore(int[] tokens, int power) {
        if (tokens.length == 0) return 0;
        
        Arrays.sort(tokens);
        int l = 0; int r = tokens.length-1;
        int currPower = power;
        int currScore = 0;
        while (l <= r) {
            
            if (currPower < tokens[l] && currScore > 0 && r-l > 1) {
                currPower += tokens[r];
                r--;
                currScore -= 1;
            } else if (currPower >= tokens[l]) {
                currPower -= tokens[l];
                l++;
                currScore += 1;
            } else {
                return currScore;
            }
        }
        
        return currScore;
        
    }
}

File: 948-bag-of-tokens/948-bag-of-tokens.py
class Solution:
    def bagOfTokensScore(self, tokens: List[int], power: int) -> int:
        if not tokens: return 0
        tokens.sort()
        # go greedy then dp
        
        l, r = 0, len(tokens)-1
        curr_power = power
        curr_score = 0
        while l <= r:
            
            if curr_power < tokens[l] and curr_score > 0 and r-l > 1:
                curr_power += tokens[r]
                r-=1
                curr_score -= 1
            elif curr_power >= tokens[l]:
                curr_power -= tokens[l]
                l+= 1
                curr_score += 1
            else:
                return curr_score
        
        return curr_score
        
        